src/audio/google/google_rtc_audio_processing.c

```diff
@@ -294,85 +277,121 @@ static int google_rtc_audio_processing_cmd_set_data(
+			 cdata->cmd);
 		return -EINVAL;
 	}
 }
 
-static int google_rtc_audio_processing_cmd_get_data(
-	struct comp_dev *dev,
-	struct sof_ipc_ctrl_data *cdata,
-	int max_data_size)
+static int google_rtc_audio_processing_cmd_get_data(struct processing_module *mod,
+						    struct sof_ipc_ctrl_data *cdata,
+						    size_t max_data_size)
 {
-	struct google_rtc_audio_processing_comp_data *cd = comp_get_drvdata(dev);
+	struct google_rtc_audio_processing_comp_data *cd = module_get_private_data(mod);
 
-	comp_info(dev, "google_rtc_audio_processing_ctrl_get_data(): %u", cdata->cmd);
+	comp_info(mod->dev, "google_rtc_audio_processing_ctrl_get_data(): %u", cdata->cmd);
 
 	switch (cdata->cmd) {
 	case SOF_CTRL_CMD_BINARY:
 		return comp_data_blob_get_cmd(cd->tuning_handler, cdata, max_data_size);
 	default:
-		comp_err(dev,
-				 "google_rtc_audio_processing_ctrl_get_data(): Only binary controls supported %d",
-				 cdata->cmd);
+		comp_err(mod->dev,
+			 "google_rtc_audio_processing_ctrl_get_data(): Only binary controls supported %d",
+			 cdata->cmd);
 		return -EINVAL;
 	}
 }
+#endif
 
-static int google_rtc_audio_processing_cmd(struct comp_dev *dev, int cmd, void *data,
-									 int max_data_size)
+static int google_rtc_audio_processing_set_config(struct processing_module *mod, uint32_t param_id,
+						  enum module_cfg_fragment_position pos,
+						  uint32_t data_offset_size,
+						  const uint8_t *fragment,
+						  size_t fragment_size, uint8_t *response,
+						  size_t response_size)
 {
-	struct sof_ipc_ctrl_data *cdata = data;
-
-	comp_dbg(dev, "google_rtc_audio_processing_cmd(): %d - data_cmd: %d", cmd, cdata->cmd);
+#if CONFIG_IPC_MAJOR_4
+	struct google_rtc_audio_processing_comp_data *cd = module_get_private_data(mod);
+	int ret;
 
-	switch (cmd) {
-	case COMP_CMD_SET_VALUE:
-	case COMP_CMD_GET_VALUE:
-		return 0;
-	case COMP_CMD_SET_DATA:
-		return google_rtc_audio_processing_cmd_set_data(dev, cdata);
-	case COMP_CMD_GET_DATA:
-		return google_rtc_audio_processing_cmd_get_data(dev, cdata, max_data_size);
-	default:
-		comp_err(dev, "google_rtc_audio_processing_cmd(): Unknown cmd %d", cmd);
+	switch (param_id) {
+	case SOF_IPC4_SWITCH_CONTROL_PARAM_ID:
+	case SOF_IPC4_ENUM_CONTROL_PARAM_ID:
+		comp_err(mod->dev, "google_rtc_audio_processing_ctrl_set_data(): Only binary controls supported");
 		return -EINVAL;
 	}
+
+	ret = comp_data_blob_set(cd->tuning_handler, pos, data_offset_size,
+				 fragment, fragment_size);
+	if (ret)
+		return ret;
+
+	/* Accept the new blob immediately so that userspace can write
+	 * the control in quick succession without error.
+	 * This ensures the last successful control write from userspace
+	 * before prepare/copy is applied.
+	 * The config blob is not referenced after reconfigure() returns
+	 * so it is safe to call comp_get_data_blob here which frees the
+	 * old blob. This assumes cmd() and prepare()/copy() cannot run
+	 * concurrently which is the case when there is no preemption.
```

Comment:
in principle `.copy()` can run concurrently with IPCs, in which case the timer interrupt can interrupt IPC processing, then the scheduler will choose between the IPC thread and the LL scheduler thread and will pick up the latter, since it has a higher priority. But if the component has only one sink and one source, mostly it's only the "stop" IPC that can arrive while the stream is already running, or can some configuration be sent to a running stream too?

Reply:
I preserved this part as close to the original so this should not change the operation. @lkoenig What do you think of this?

Reply:
Configurations can be sent when a stream is running.

Reply:
@lyakh @lkoenig I think it depends on the (unknown) internals of the processing whether new config data blob during copy() is a hazard  or not. The open-source mock-up version ignores the blob content totally.

The reconfigure set here should ensure that the next copy() correctly re-initializes for the blob. Please let me know if this needs to be addressed now, I could propose a fix. Or (preferably) let @thesofproject/google  address this in successive PR.

Reply:
The data should be whole and not altered during the call of `GoogleRtcAudioProcessingReconfigure` and `GoogleRtcAudioProcessingParseSofConfigMessage`
As of today we do not have concurrent access protection enabled in the `GoogleRtcAudioProcessing`.

Reply:
@singalsu at least I would note in the comment, that a race is possible and should be fixed or at least further investigated